### PR TITLE
IGNITE-21119 Complete index creation future when the index is AVAILABLE or removed

### PR DESCRIPTION
**Pull request by rpuch**
https://issues.apache.org/jira/browse/IGNITE-21119

### CODE REVIEW

File: modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItBuildIndexOneNodeTest.java [hunk 1]
```diff
@@ -93,7 +87,7 @@ void testRecoverBuildingIndex() throws Exception {
             return false;
         });
 
-        createIndex(TABLE_NAME, INDEX_NAME, "ID");
+        runAsync(() -> createIndex(TABLE_NAME, INDEX_NAME, "ID"));
```
Comment (tkalkirill) [inline modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItBuildIndexOneNodeTest.java:90]: Maybe it would be better to wait for the futures to finish executing?

File: modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItBuildIndexOneNodeTest.java [hunk 2]
```diff
@@ -285,13 +279,10 @@ void testBuildingIndexWithUpdateSchemaAfterCreateIndex() throws Exception {
         Transaction rwTx = node().transactions().begin(new TransactionOptions().readOnly(false));
 
         try {
-            setAwaitIndexAvailability(false);
-
-            createIndex(TABLE_NAME, INDEX_NAME, columName);
+            runAsync(() -> createIndex(TABLE_NAME, INDEX_NAME, columName));
```
Comment (tkalkirill) [inline modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItBuildIndexOneNodeTest.java:282]: Same

File: modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItDropIndexMultipleNodesTest.java [hunk 3]
```diff
@@ -239,7 +242,12 @@ void testDropIndexAfterRegistering() {
         runInRwTransaction(CLUSTER.aliveNode(), tx -> {
             // Create an index inside a transaction, this will prevent the index from building.
             try {
-                createIndex();
+                runAsync(() -> createIndex());
```
Comment (tkalkirill) [inline modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItDropIndexMultipleNodesTest.java:245]: Same about wait future.

File: modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItDropIndexMultipleNodesTest.java [hunk 4]
```diff
@@ -264,7 +269,7 @@ void testDropIndexDuringBuilding() {
 
         CompletableFuture<Void> indexRemovedFuture = indexRemovedFuture();
 
-        createIndex();
+        runAsync(() -> createIndex());
```
Comment (tkalkirill) [inline modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItDropIndexMultipleNodesTest.java:272]: Same

File: modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItDropIndexOneNodeTest.java [hunk 5]
```diff
@@ -65,14 +69,23 @@ void testCreateIndexAfterDrop() {
...
     void testCreateIndexAfterDropWhileTransactionInProgress() {
+        CatalogManager catalogManager = CLUSTER.aliveNode().catalogManager();
+
         runInRwTransaction(tx -> {
             dropIndex(INDEX_NAME);
 
-            // The new index will not become available, since we are inside a transaction that has been started before this index was
-            // created.
-            setAwaitIndexAvailability(false);
+            runAsync(() -> createIndex(TABLE_NAME, INDEX_NAME, COLUMN_NAME));
```
Comment (tkalkirill) [inline modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItDropIndexOneNodeTest.java:77]: same

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogIndexStatus.java [hunk 6]
```diff
@@ -80,6 +80,19 @@ public boolean isAlive() {
         return this != STOPPING;
     }
 
+    /**
+     * Returns {@code true} if the index is {@link #AVAILABLE} or follows it.
+     */
+    public boolean isAvailableOrLater() {
```
Comment (tkalkirill) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogIndexStatus.java:86]: Maybe move to `CatalogUtils` ?

File: modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogManagerSelfTest.java [hunk 7]
```diff
@@ -217,6 +217,22 @@ public void testEmptyCatalog() {
         assertThat(manager.latestCatalogVersion(), is(0));
     }
 
+    @Test
+    public void assignsSuccessiveCatalogVersions() {
+        CompletableFuture<Integer> version1Future = manager.execute(simpleTable(TABLE_NAME));
+        assertThat(version1Future, willCompleteSuccessfully());
```
Comment (tkalkirill) [inline modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogManagerSelfTest.java:223]: Can you do `willBe(1)` and etc?

File: modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItDropIndexOneNodeTest.java [hunk 8]
```diff
@@ -65,14 +69,23 @@ void testCreateIndexAfterDrop() {
...
+
+            try {
+                assertTrue(waitForCondition(
```
Comment (tkalkirill) [inline modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItDropIndexOneNodeTest.java:80]: Let's add it as a helper method to the test code?

File: modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItRwTransactionAndIndexesTest.java [hunk 9]
```diff
@@ -62,16 +66,23 @@ void tearDown() {
...
+
         TableImpl table = (TableImpl) createZoneAndTable(ZONE_NAME, TABLE_NAME, 1, 1, ENGINE_NAME);
 
-        setAwaitIndexAvailability(false);
         dropAnyBuildIndexMessages();
 
         Transaction rwTx = beginRwTransaction();
 
-        createIndex(TABLE_NAME, INDEX_NAME, COLUMN_NAME);
+        runAsync(() -> createIndex(TABLE_NAME, INDEX_NAME, COLUMN_NAME));
```
Comment (tkalkirill) [inline modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItRwTransactionAndIndexesTest.java:80]: Same

File: modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItRwTransactionAndIndexesTest.java [hunk 10]
```diff
@@ -62,16 +66,23 @@ void tearDown() {
...
+
+        assertTrue(waitForCondition(
```
Comment (tkalkirill) [inline modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItRwTransactionAndIndexesTest.java:82]: Same

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/ItDataSchemaSyncTest.java [hunk 11]
```diff
@@ -198,18 +198,14 @@ public void queryWaitAppropriateSchema() {
 
         createTable(ignite0, TABLE_NAME);
 
-        WatchListenerInhibitor listenerInhibitor = WatchListenerInhibitor.metastorageEventsInhibitor(ignite1);
-
-        listenerInhibitor.startInhibit();
+        WatchListenerInhibitor.metastorageEventsInhibitor(ignite1).withInhibition(() -> {
+            runAsync(() -> sql(ignite0, "CREATE INDEX idx1 ON " + TABLE_NAME + "(valint)"));
```
Comment (tkalkirill) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/ItDataSchemaSyncTest.java:202]: same

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/ItTablesApiTest.java [hunk 12]
```diff
@@ -256,6 +263,13 @@ public void testAddIndex() {
         tryToCreateIndex(ignite0, TABLE_NAME, false);
     }
 
+    private static void waitForIndexToAppearInAnyState(IgniteImpl ignite0) throws InterruptedException {
```
Comment (tkalkirill) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/ItTablesApiTest.java:266]: same

File: modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItIndexDdlTest.java [hunk 13]
```diff
@@ -83,10 +95,12 @@ void testDropIndex() {
      * @param failIfExist Throw an exception if the index exist.
      */
     private static void tryToCreateIndex(String tableName, String indexName, boolean failIfExist) {
+        Loggers.forClass(ItIndexDdlTest.class).info("XXX starting index creation");
```
Comment (tkalkirill) [inline modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItIndexDdlTest.java:98]: Let's move the logger to the class and not use "XXX".

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ddl/DdlCommandHandler.java [hunk 14]
```diff
@@ -169,9 +191,74 @@ private static BiFunction<Object, Throwable, Boolean> handleModificationResult(b
...
     }
 
+    private CompletionStage<Void> waitTillIndexBecomesAvailableOrRemoved(CreateIndexCommand cmd, Integer creationCatalogVersion) {
+        CompletableFuture<Void> future = new CompletableFuture<>();
+
+        Catalog catalog = catalogManager.catalog(creationCatalogVersion);
+        assert catalog != null : creationCatalogVersion;
+
+        CatalogSchemaDescriptor schema = catalog.schema(cmd.schemaName());
+        assert schema != null : cmd.schemaName();
```
Comment (tkalkirill) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ddl/DdlCommandHandler.java:205]: I think it would be good to indicate the catalog version in the error message.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ddl/DdlCommandHandler.java [hunk 15]
```diff
@@ -169,9 +191,74 @@ private static BiFunction<Object, Throwable, Boolean> handleModificationResult(b
...
+
+        CatalogIndexDescriptor index = schema.aliveIndex(cmd.indexName());
+        assert index != null : cmd.indexName();
```
Comment (tkalkirill) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ddl/DdlCommandHandler.java:208]: same

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ddl/DdlCommandHandler.java [hunk 16]
```diff
@@ -169,9 +191,74 @@ private static BiFunction<Object, Throwable, Boolean> handleModificationResult(b
...
+
+        EventListener<CatalogEventParameters> availabilityListener = event -> {
+            if (((MakeIndexAvailableEventParameters) event).indexId() == index.id()) {
```
Comment (tkalkirill) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ddl/DdlCommandHandler.java:211]: U can use `EventListener#fromConsumer`

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ddl/DdlCommandHandler.java [hunk 17]
```diff
@@ -169,9 +191,74 @@ private static BiFunction<Object, Throwable, Boolean> handleModificationResult(b
...
+                completeFutureWhenEventVersionActivates(future, event);
+            }
+
+            return falseCompletedFuture();
+        };
+        catalogManager.listen(CatalogEvent.INDEX_AVAILABLE, availabilityListener);
+
+        EventListener<CatalogEventParameters> removalListener = event -> {
+            if (((RemoveIndexEventParameters) event).indexId() == index.id()) {
```
Comment (tkalkirill) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ddl/DdlCommandHandler.java:220]: Same about fromComsumer

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogManager.java [hunk 18]
```diff
@@ -29,16 +29,16 @@ public interface CatalogManager extends IgniteComponent, CatalogService {
      * Executes given command.
      *
      * @param command Command to execute.
-     * @return Future representing result of execution.
+     * @return Future representing result of execution (it will be completed with the created catalog version).
      */
-    CompletableFuture<Void> execute(CatalogCommand command);
+    CompletableFuture<Integer> execute(CatalogCommand command);
```
Comment (tkalkirill) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogManager.java:34]: Didn’t notice where you are using the result of the future, not in the test code?

File: modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogManagerSelfTest.java (hunk 7 above)
Comment (rpuch) [reply modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogManagerSelfTest.java:223]: There is a test that checks explicitly that `execute()` returns successive version numbers. Non of other tests actually needs to know the completion value, it's enough to verify that the future completes with any value. If I add that, this will make the test very rigid for no reason.

File: modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItRwTransactionAndIndexesTest.java (hunk 9 above)
Comment (rpuch) [reply modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItRwTransactionAndIndexesTest.java:80]: This test is built in such a way that the iindex never becomes AVAILABLE

File: modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItRwTransactionAndIndexesTest.java (hunk 10 above)
Comment (rpuch) [reply modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItRwTransactionAndIndexesTest.java:82]: But it's the same test as above. If the concern is different, please elaborate.

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogManager.java (hunk 18 above)
Comment (rpuch) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogManager.java:34]: In `DdlCommandHandler#handleCreateIndex()`

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogIndexStatus.java (hunk 6 above)
Comment (rpuch) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogIndexStatus.java:86]: It's about the index status, so it's naturally to have it on it.

File: modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItRwTransactionAndIndexesTest.java (hunk 10 above)
Comment (rpuch) [reply modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItRwTransactionAndIndexesTest.java]: Oh, I see what you mean. Fixed.

**Review by tkalkirill — APPROVED** [review]

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ddl/DdlCommandHandler.java [hunk 19]
```diff
@@ -169,12 +200,86 @@ private static BiFunction<Object, Throwable, Boolean> handleModificationResult(b
...
+                future.complete(null);
+            }
+        });
+        catalogManager.listen(CatalogEvent.INDEX_REMOVED, removalListener);
+
+        // We added listeners, but the index could switch to a state of interest before we added them, so check
+        // explicitly.
+        int latestVersion = catalogManager.latestCatalogVersion();
+        for (int version = creationCatalogVersion + 1; version <= latestVersion; version++) {
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ddl/DdlCommandHandler.java:237]: why do we iterate over all future versions? Is it not enough to check the latest one?

Comment (rpuch) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ddl/DdlCommandHandler.java:237]: If we missed a few catalog updates, this allows us to find the earliest of them in which the index has become AVAILABLE, so the user will need to wait a little less (because we don't just complete the future as soon as we see an AVAILABLE version in the Catalog, we take its activation ts and wait till it comes). Also, it's unlikely that a thousand catalog versions will appear for scanning while we are adding the listeners :)

To sum up: the scanning might yield an earlier completion moment, but if we take the latest version, it won't be too different, so I'll change this if you insist.

Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ddl/DdlCommandHandler.java:237]: > the scanning might yield an earlier completion moment

I missed this part, thanks for explanation! I'm ok with current approach

File: modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItIndexDdlTest.java [hunk 20]
```diff
@@ -98,4 +114,52 @@ private static void tryToCreateIndex(String tableName, String indexName, boolean
     private static void tryToDropIndex(String indexName, boolean failIfNotExist) {
         sql(String.format("DROP INDEX %s", failIfNotExist ? indexName : "IF EXISTS " + indexName));
     }
+
+    private static <T> T preventingIndexBuild(Supplier<T> supplier) {
+        return CLUSTER.aliveNode().transactions().runInTransaction((Function<Transaction, T>) tx -> supplier.get());
+    }
+
+    @Test
+    public void createIndexFutureCompletesWhenIndexBecomesAvailable() {
```
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItIndexDdlTest.java:123]: I added another test with production-default cluster configuration (including DD=1s) that uses an index right after the DDL returns.

**Review by korlov42 — APPROVED** [review]
